gluten-core/src/main/java/org/apache/gluten/memory/memtarget/spark/DynamicOffHeapSizingPolicyChecker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.memory.memtarget.spark;
+
+import org.apache.gluten.GlutenConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class DynamicOffHeapSizingPolicyChecker {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DynamicOffHeapSizingPolicyChecker.class);
+  private final long maxOnHeapMemoryInBytes = GlutenConfig.getConf().onHeapMemorySize();
+  private final AtomicLong usedOffHeapBytes = new AtomicLong();
+
+  DynamicOffHeapSizingPolicyChecker() {}
+
+  public boolean canBorrow(long size) {
```

Comment:
This code says when native requires memory, we borrow onheap memory to it.  But then later onheap requires memory, how can onheap get the actually memory before native repay? I think It would be killed by yarn or k8s due to it uses double memory at this time. And It's a very general case, e.g., there are stages running in parallel or there are many queries running in parallel.

Reply:
I believe you are referring to the caveat raised above that "it is possible that the Java codebase can oversubscribe memory as it allocates, which is under the on-heap quota, although there is enough off-heap usage where we should fail the allocation". Feedback was also given above that this should be documented in the code as well to be very explicit about the risk. I will do that.

Reply:
Oversubscribe memory is quite risk. I think we should at least respect the Spark idea. The max execution memory in Spark is:
```
(spark.executor.memory - 300MB) * spark.memory.fraction
```

`totalMemory() - freeMemory()` breaks this idea. Even if the execution offload to native, there are still many things depend on on-heap memory, e.g., the metadata(plan, split, map status, event...), broadcast block, rdd cache. One thought is that, we'd better to follow the Spark, and change it to:
```
(totalMemory() - freeMemory() - 300MB) * spark.memory.fraction
```

Reply:
Agreed. With the proposed change, we are seeing the "max on heap" as being the spark.executor.memory setting, but that does go against Spark's setting aside memory accordingly. I'll change the proposed logic to follow what Spark has determined as the max on-heap.